src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -348,6 +350,28 @@ static inline int MPIDI_OFI_do_am_isend(int rank,
     mpi_errno = MPIDI_OFI_am_init_request(am_hdr, am_hdr_sz, sreq);
     MPIR_ERR_CHECK(mpi_errno);
 
+    MPL_pointer_attr_t attr;
+    MPL_gpu_query_pointer_attr(buf, &attr);
+
+    if (data_sz && attr.type == MPL_GPU_POINTER_DEV) {
+        /* If dt_contig and HMEM is enabled no need to stage GPU data in
+         * host memory. libfabric can handle directly the GPU using IOV
+         * in fi_sendv. */
+        if (!(dt_contig && MPIDI_OFI_ENABLE_HMEM)) {
```

Comment:
Not all noncontiguous data needs packing.  It needs to be separated out based on density.

Reply:
I think at this moment we only allow for one data segment in iov as this is statically allocated with size `3`. This should be probably revised in a separate PR and make iov dynamically allocated to allow for more data segments.

Reply:
This IOV is not the same as the IOV used for datatype communication.  All of the actual user data is only represented by the datatype.  The IOV here is only to provide different header segments.

Reply:
I am afraid I don't follow. The first two iov elements are for the header, the third is for the buffer right? we always do packing for non contig buffers and pass the packed buffer as 3rd iov element. for high density types we can let libfabric take an iov in which the number of elements is iov_len from the datatype itself + 2 for the header segments. I might be missing something here.

Reply:
> we always do packing for non contig buffers and pass the packed buffer as 3rd iov element

I don't think so.  In fact, I don't think we ever use the third element of the IOV vector.  We use the first two elements for the header and extended header.  The data buffer (which can be noncontiguous) is passed down separately from the ch4 layer.  The data buffer is not a contiguous segment.  It could be a pointer to a noncontiguous buffer. 
 This IOV is not meant to pass contiguous segments.  In fact, this is exactly the reason why I was suggesting that this function should not take a IOV at all.  It should take two parameters for the two headers.